onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -71,22 +68,54 @@ CUDAExecutionProvider::PerThreadContext::PerThreadContext(OrtDevice::DeviceId de
   allocator_ = CreateAllocator(default_memory_info, device_id);
 }
 
+cublasHandle_t CUDAExecutionProvider::PerThreadContext::CublasHandle() {
```

Comment:
Changing these PerThreadContext CUDA handles to be created lazily because there was a segmentation fault during handle cleanup in one of the Python training tests.

In particular, onnxruntime_test_ort_trainer.py:TestOrtTrainer.testBertCheckpointingBasic creates a TrainingSession (and hence a CUDAExecutionProvider) and never calls InferenceSession::Run(). This leads to the handle cleanup happening fairly late, later than the destruction of the CUDAExecutionProvider which is where it normally is cleaned up if Run() is called. In that case, a segmentation fault happens when calling cudnnDestroy().

This avoids the problem by only creating/destroying the cuDNN handle (and other ones) if needed.

Reply:
This change looks good to me.

---
In reply to: [424816330](https://github.com/microsoft/onnxruntime/pull/3767#discussion_r424816330) [](ancestors = 424816330)

Reply:
Sorry I don't understand. Why it would crash? When it calls cublasDestroy(cublas_handle_), it doesn't depend on ORT?

Reply:
I don't understand why it crashed either. But it was a segmentation fault from the call to cudnnDestroy().

---
In reply to: [425946928](https://github.com/microsoft/onnxruntime/pull/3767#discussion_r425946928) [](ancestors = 425946928)

Reply:
Will you have sometime to investigate it afterwards? Or at least having a unitest that can catch the crash. Otherwise nobody can change this code anymore.

Reply:
will look into it some more after this PR.